packages/astro/src/core/app/index.ts

```diff
@@ -280,12 +282,21 @@ export class App {
 					status
 				);
 				const page = (await mod.page()) as any;
-				if (mod.onRequest) {
+				if (runMiddleware && mod.onRequest) {
 					this.#pipeline.setMiddlewareFunction(mod.onRequest as MiddlewareEndpointHandler);
+					usedMiddleware = true;
+				}
+				if (!runMiddleware) {
+					this.#pipeline.unsetMiddlewareFunction();
 				}
 				const response = await this.#pipeline.renderRoute(newRenderContext, page);
 				return this.#mergeResponses(response, originalResponse);
-			} catch {}
+			} catch {
+				// Middleware may be the cause of the error, so we try rendering 404/500.astro without it.
+				if (runMiddleware && usedMiddleware) {
```

Comment:
We can remove `usedMiddleware` and use `mod.onRequest`. Or have the `pipeline` give us this information.